File: Decision Tree Regression/Decision_Tree_Regression/decision-tree_regression1.py
# -*- coding: utf-8 -*-
"""
Created on Sun Dec 24 17:53:50 2017

@author: Zeeshan K
"""

# Descion Tree Regression Model


# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
# Just to check weather the employee is telling truth about his/her salary or bluffing
X = dataset.iloc[:, 1:2].values # transforming the column into matrix of feature so that we can apply polynomial rules
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""


# Fitting Descion Tree Model to the Dataset
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X, y)



#predict a new result with Decision Tree Regression
y_pred = regressor.predict(6.5)

# Visualising the Descion Tree Regression results
plt.scatter(X, y, color = 'black')
plt.plot(X, regressor.predict(X), color = 'purple')
plt.title("Truth or Bluff (Decision Tree Regression)")
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()


#Visualising the Descion Tree Regression result for higher resolution and smoother curve
X_grid = np.arange(min(X), max(X), 0.01)
X_grid = X_grid.reshape((len(X_grid),1))
plt.scatter(X, y, color = 'black')
plt.plot(X_grid, regressor.predict(X_grid), color = 'purple')
plt.title("Truth or Bluff (Ploynomial Regression)")
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

File: Decision Tree Regression/Decision_Tree_Regression/decision_tree_regression1.R
#Decision tree Regression 

# Data Preprocessing 

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3] # select only level and salary column

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
#library(caTools)
#set.seed(123)
#split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
#training_set = subset(dataset, split == TRUE)
#test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)


# Fitting dataset into Decision Tree Model
#install.packages('rpart')
library(rpart)
regressor = rpart(formula = Salary ~ .,
                  data = dataset,
                  control = rpart.control(minsplit = 1))

# predict the new result with Decision Tree Regression
y_pred = predict(regressor, data.frame(Level= 6.5))


# Visualising the Decision Tree Regression result
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level),0.01)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color ='red') + 
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level = x_grid))),
            color ='blue') +
  ggtitle('Truth or Bluff (Decision Tree Regression Model)') +
  xlab('Levels') +
  ylab('Salary') 
##### shows a problem in graph in a straight line, problem is related to splits, and we have depend variable which 
# is multiples of hundereds and needs the intervals by giving the conditions on splits, for this we can give parameter of condition to model
# rpart.control(minsplit = 1),  after which it also take us to trap, red flag, as it gives more than one splits here and 
# Decision tree splits the independent variables into several intervals, which took the avg of dependent variables but it
# is not splits the intervals properly, so we need to check graphs resolution ####


# Preidict the new result with Regression
y_pred_poly = predict(regressor, data.frame(Level =6.5, 
                                            Level2 =6.5^2, 
                                            Level3 = 6.5^3, 
                                            Level4 = 6.5^4))
